firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/Firestore.kt

```diff
@@ -163,3 +173,51 @@ class FirebaseFirestoreKtxRegistrar : ComponentRegistrar {
     override fun getComponents(): List<Component<*>> =
             listOf(LibraryVersionComponent.create(LIBRARY_NAME, BuildConfig.VERSION_NAME))
 }
+
+/**
+ * Attach a snapshotListener to a DocumentReference and use it as a coroutine flow
+ * @param metadataChanges Indicates whether metadata-only changes
+ */
+fun DocumentReference.toFlow(metadataChanges: MetadataChanges = MetadataChanges.EXCLUDE) = flow {
+    val channel = Channel<DocumentSnapshot>(Channel.CONFLATED)
+    var listener: ListenerRegistration? = null
+    withContext(Dispatchers.Main.immediate) {
+        listener = addSnapshotListener(metadataChanges) { value, error ->
+            value?.let { channel.offer(it) }
+            error?.let { channel.close(it) }
+        }
+    }
+    try {
+        for (value in channel) {
+            emit(value)
+        }
+    } finally {
+        withContext(Dispatchers.Main.immediate + NonCancellable) {
+            listener?.remove()
+        }
+    }
+}
+
+/**
+ * Attach a snapshotListener to a Query and use it as a coroutine flow
+ * @param metadataChanges Indicates whether metadata-only changes
+ */
+fun Query.toFlow(metadataChanges: MetadataChanges = MetadataChanges.EXCLUDE): Flow<QuerySnapshot> = flow {
+    val channel = Channel<QuerySnapshot>(Channel.CONFLATED)
+    var listener: ListenerRegistration? = null
+    withContext(Dispatchers.Main.immediate) {
+        listener = addSnapshotListener(metadataChanges) { value, error ->
+            value?.let { channel.offer(it) }
+            error?.let { channel.close(it) }
+        }
+    }
+    try {
```

Comment:
```suggestion
    try {
        withContext(Dispatchers.Main.immediate) {
            listener = addSnapshotListener(metadataChanges) { value, error ->
                value?.let { channel.offer(it) }
                error?.let { channel.close(it) }
            }
        }
```
Fixes a leak of the listener that would occur in case of cancellation of the coroutine on exiting the `withContext` block.